### PR TITLE
Fix all clippy warnings, more idiomatic code

### PR DESCRIPTION
**Pull request by rsalmei**
Hi,
I'm Rogerio from linkedin, I've interacted with you there.
Here is my contribution, ~~most~~ all clippy warnings are fixed, the code is restructured to be easier to use, and refactored to be more idiomatic. I hope you like it.

EDIT: I just fixed the last clippy warnings!

EDIT2: You need TESTS! The process of refactoring code and manually testing it for all cases can be quite annoying. This should be your first task now, create tests to ensure all combinations of input parameters do work.

EDIT3: The next task you should tackle is to improve the error handling. `unwrap()` and `expect()` should not ever be used on production code. Even on CLI tools like this one, panicking is embarrassing. All functions that can fail should return either Result or Option and be handled appropriately. Use `?` when the proper handler is the caller. I've made `main() -> Result<(), Box<dyn Error>>` now, so anything that does not have a default or can't be handled, just `?` its way to the main.

### CODE REVIEW

**Comment by mainak55512** [comment]
Hi Rogerio @rsalmei ,

Glad to see you! 👋
The refactor is excellent and the code is running even faster! so thanks for that.
Just one thing, I have checked out the code locally and ran, it seems there is a small issue, when comparing two string values it is giving false.

Actual JSON input:
image

Output of query involving string comparison:
image

Output of query involving numbers only:
image

I am not sure if I am doing something wrong, but could you please check and let me know if there is any issue at all, and if is, could you please fix that as well?

Input file: 
[data.json](https://github.com/user-attachments/files/17375420/data.json)

File: src/query/interpreter.rs [hunk 1]
```diff
@@ -47,77 +56,69 @@ fn _eval_binary_expr(
...
                     .parse::<f64>()
                     .expect("Not a Number")
                     == right.parse::<f64>().expect("Not a Number"),
             ),
-            _ => {
-                return RuntimeType::Bool(
-                    get_value_from_obj(obj, &left).to_string() == right.to_string(),
-                );
-            }
+            _ => RuntimeType::Bool(get_value_from_obj(obj, left) == right),
```
Comment (mainak55512) [inline src/query/interpreter.rs:71]: Changing this to 
_ => RuntimeType::Bool(get_value_from_obj(obj, left).to_string() == right.to_string()),
fixes the issue.

File: src/query/interpreter.rs [hunk 2]
```diff
@@ -47,77 +56,69 @@ fn _eval_binary_expr(
...
                     .parse::<f64>()
                     .expect("Not a Number")
                     != right.parse::<f64>().expect("Not a Number"),
             ),
-            _ => {
-                return RuntimeType::Bool(
-                    get_value_from_obj(obj, &left).to_string() != right.to_string(),
-                )
-            }
+            _ => RuntimeType::Bool(get_value_from_obj(obj, left) != right),
```
Comment (mainak55512) [inline src/query/interpreter.rs:121]: Changing this to
_ => RuntimeType::Bool(get_value_from_obj(obj, left).to_string() == right.to_string()),
fixes the issue.

**Comment by rsalmei** [comment]
Hey @mainak55512, that's awesome, you're welcome!

Regarding the issue of comparing texts, that's weird. It's the last set of clippy issues I fixed, in the last commit:
[fix last clippy warning (serde_json::Value has a PartialEQ<&str> impl)](https://github.com/mainak55512/rjq/pull/10/commits/a8672cb9e2552b74b798836102c78050e4e0af16)

I'll try to see what's happening later today.

**Comment by mainak55512** [comment]
> Hey @mainak55512, that's awesome, you're welcome!
> 
> Regarding the issue of comparing texts, that's weird. It's the last set of clippy issues I fixed, in the last commit:
> [fix last clippy warning (serde_json::Value has a PartialEQ<&str> impl)](https://github.com/mainak55512/rjq/pull/10/commits/a8672cb9e2552b74b798836102c78050e4e0af16)
> 
> I'll try to see what's happening later today.

Thank you @rsalmei ,

I have merged the pull request, really appreciate your efforts!
For now, for that issue I converted both sides to string and compared. If there is a better approach you can find feel free to create new pull requests... ☺️

**Comment by rsalmei** [comment]
I'm glad you liked. I love Rust, as well as spreading the word about how to make better use of it.

I found it does work, only it doesn't need the single quotes now:

```
> cargo run -- < ~/Desktop/data.json --query="name = Test"
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/rjq '--query=name = Test'`
[
  {
    "age": 25,
    "name": "Test"
  },
  {
    "age": 19,
    "name": "Test"
  }
]
```

**Comment by mainak55512** [comment]
> I'm glad you liked. I love Rust, as well as spreading the word about how to make better use of it.
> 
> I found it does work, only it doesn't need the single quotes now:
> 
> ```
> > cargo run -- < ~/Desktop/data.json --query="name = Test"
>     Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
>      Running `target/debug/rjq '--query=name = Test'`
> [
>   {
>     "age": 25,
>     "name": "Test"
>   },
>   {
>     "age": 19,
>     "name": "Test"
>   }
> ]
> ```

Hi @rsalmei , thanks for digging that out 👍, but it didn't work for the release build when I earlier tried this way(without quotes). But in my opinion, the strings should be wrapped in quotes, let me know your opinion as well 🤝

**Pull request merged**